auth/auth.go

```diff
@@ -309,39 +309,62 @@ func (auth *Authenticator) UpdateSequenceNumber(p Principal, seq uint64) error {
 }
 
 // Invalidates the channel list of a user/role by saving its Channels() property as nil.
-func (auth *Authenticator) InvalidateChannels(p Principal, invalSeq uint64) error {
-	invalidateChannelsCallback := func(p Principal) (updatedPrincipal Principal, err error) {
+func (auth *Authenticator) InvalidateChannels(name string, isUser bool, invalSeq uint64) error {
+	for i := 1; i < PrincipalUpdateMaxCasRetries; i++ {
+		var princ Principal
+		var docID string
+
+		if isUser {
+			princ = &userImpl{}
+			docID = docIDForUser(name)
+		} else {
+			princ = &roleImpl{}
+			docID = docIDForRole(name)
+		}
 
-		if p == nil || p.Channels() == nil {
-			return p, base.ErrUpdateCancel
+		cas, err := auth.bucket.Get(docID, &princ)
+		if err != nil {
+			return err
 		}
 
-		base.Infof(base.KeyAccess, "Invalidate access of %q", base.UD(p.Name()))
+		princ.SetChannelInvalSeq(invalSeq)
 
-		p.SetChannelInvalSeq(invalSeq)
-		return p, nil
+		cas, err = auth.bucket.WriteCas(docID, 0, 0, cas, princ, 0)
+		if err == nil {
+			return nil
+		}
+
+		if !base.IsCasMismatch(err) {
+			return err
+		}
 	}
-	return auth.casUpdatePrincipal(p, invalidateChannelsCallback)
+	return nil
+
 }
 
 // Invalidates the role list of a user by saving its Roles() property as nil.
-func (auth *Authenticator) InvalidateRoles(user User, invalSeq uint64) error {
+func (auth *Authenticator) InvalidateRoles(user string, invalSeq uint64) error {
```

Comment:
We haven't discussed previously, but if a single document triggers invalidation of both roles and channels for a user, can we do that in a single operation?

Reply:
It might be possible. Will look into it

Reply:
There's quite a lot of work around where we check for changed principals so I think it'd be a bigger piece of work. We could possibly flag this for future improvement?

Reply:
Makes sense.